File: sdk/gapura/python/src/hara_gapura/identity.py
"""Identity resource: resolve DIDs, register, bindings."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from .types import (
    Binding,
    DidResolution,
    IdentityRegistration,
    IdentityRequest,
    Operation,
    SubjectType,
)

if TYPE_CHECKING:
    from .client import GapuraClient

__all__ = ["Identity"]


class Identity:
    """``client.identity`` — resolve/register DIDs and manage Authentik bindings."""

    def __init__(self, client: "GapuraClient") -> None:
        self._client = client

    def resolve(self, did: str) -> DidResolution:
        """Resolve a ``did:hara`` DID (``GET /did/{did}``, scope ``identity:read``)."""
        return self._client._request("GET", f"did/{quote(did, safe='')}")

    def register(
        self,
        *,
        subject_type: SubjectType,
        display_name: str,
        controller_jwk: dict[str, Any] | None = None,
        namespace_hint: str | None = None,
    ) -> IdentityRegistration | Operation:
        """Issue/register a DID (``POST /identities``, scope ``identity:write``).

        Returns an :class:`IdentityRegistration` on ``201`` (confirmed) or an
        :class:`Operation` on ``202`` (Sidetree batching) — poll
        :meth:`operation` for the batched case.
        """
        body: IdentityRequest = {
            "subjectType": subject_type,
            "displayName": display_name,
        }
        if controller_jwk is not None:
            body["controllerJwk"] = controller_jwk
        if namespace_hint is not None:
            body["namespaceHint"] = namespace_hint
        return self._client._request("POST", "identities", json=body)

    def operation(self, op_id: str) -> Operation:
        """Poll a Sidetree-batched registration (``GET /identities/operations/{id}``)."""
        return self._client._request(
            "GET", f"identities/operations/{quote(op_id, safe='')}"
        )

    def binding_by_subject(self, sub: str) -> Binding:
        """DID bound to an Authentik subject (``GET /identities/by-subject/{sub}``)."""
        return self._client._request(
            "GET", f"identities/by-subject/{quote(sub, safe='')}"
        )

    def create_binding(
        self, did: str, authentik_sub: str, tenant: str
    ) -> Binding:
        """Bind a DID to an Authentik subject (``POST /identities/{did}/bindings``)."""
        return self._client._request(
            "POST",
            f"identities/{quote(did, safe='')}/bindings",
            json={"authentikSub": authentik_sub, "tenant": tenant},
        )

File: services/gapura-gateway/src/routes/identity.ts
/**
 * Identity routes (§4). Resolve/register proxy to HaraDID/Numira; bindings via
 * the store. REAL wiring where a resolver URL is configured; STUB DID docs in
 * dev (see clients/haradid.ts).
 *   GET  /v1/did/:did                          scope identity:read
 *   POST /v1/identities                        scope identity:write
 *   GET  /v1/identities/operations/:id         scope identity:read
 *   GET  /v1/identities/by-subject/:sub        scope identity:read
 *   POST /v1/identities/:did/bindings          scope identity:write
 */

import type { FastifyInstance } from "fastify";
import { requireScope } from "../auth.js";
import { ProblemError, sendProblem } from "../errors.js";
import { DidClientError, HaraDidClient } from "../clients/haradid.js";
import type { Store } from "../store/index.js";
import type { IdentityRequest, Operation } from "../types.js";

const DID_RE = /^did:hara:/;

export function registerIdentityRoutes(
  app: FastifyInstance,
  deps: { store: Store; haradid: HaraDidClient },
): void {
  const { store, haradid } = deps;

  // ── GET /v1/did/:did ────────────────────────────────────────────────────────
  app.get("/v1/did/:did", { preHandler: requireScope("identity:read") }, async (req, reply) => {
    const { did } = req.params as { did: string };
    if (!DID_RE.test(did)) {
      return sendProblem(reply, "invalid_did", "not a did:hara identifier");
    }
    try {
      return reply.send(await haradid.resolve(did));
    } catch (e) {
      if (e instanceof DidClientError) {
        if (e.reason === "did_not_found") return sendProblem(reply, "did_not_found", did);
        if (e.reason === "did_deactivated") return sendProblem(reply, "did_deactivated", did);
        return sendProblem(reply, "invalid_did", e.message);
      }
      throw e;
    }
  });

  // ── POST /v1/identities ─────────────────────────────────────────────────────
  app.post("/v1/identities", { preHandler: requireScope("identity:write") }, async (req, reply) => {
    const body = req.body as IdentityRequest;
    if (!body || typeof body.displayName !== "string" || !body.subjectType) {
      return sendProblem(reply, "invalid_request", "subjectType and displayName are required");
    }
    const result = await haradid.register(body);
    // Sidetree-batched actor DIDs come back as an Operation (202); others 201.
    const isBatching = (result as Operation).state === "batching";
    return reply.code(isBatching ? 202 : 201).send(result);
  });

  // ── GET /v1/identities/operations/:operationId ──────────────────────────────
  app.get(
    "/v1/identities/operations/:operationId",
    { preHandler: requireScope("identity:read") },
    async (req, reply) => {
      const { operationId } = req.params as { operationId: string };
      return reply.send(await haradid.getOperation(operationId));
    },
  );

  // ── GET /v1/identities/by-subject/:authentikSub ─────────────────────────────
  app.get(
    "/v1/identities/by-subject/:authentikSub",
    { preHandler: requireScope("identity:read") },
    async (req, reply) => {
      const { authentikSub } = req.params as { authentikSub: string };
      const binding = await store.getBindingBySubject(authentikSub);
      if (!binding) throw new ProblemError("did_not_found", `no binding for subject ${authentikSub}`);
      return reply.send(binding);
    },
  );

  // ── POST /v1/identities/:did/bindings ───────────────────────────────────────
  app.post(
    "/v1/identities/:did/bindings",
    { preHandler: requireScope("identity:write") },
    async (req, reply) => {
      const { did } = req.params as { did: string };
      if (!DID_RE.test(did)) {
        return sendProblem(reply, "invalid_did", "not a did:hara identifier");
      }
      const body = (req.body ?? {}) as { authentikSub?: string; tenant?: string };
      if (!body.authentikSub || !body.tenant) {
        return sendProblem(reply, "invalid_request", "authentikSub and tenant are required");
      }
      const binding = await store.putBinding({
        authentikSub: body.authentikSub,
        did,
        tenant: body.tenant,
      });
      return reply.code(201).send(binding);
    },
  );
}
